### PR TITLE
sqllogictest: plumb tracing args to orchestrator

### PR DESCRIPTION
**Pull request by mjibson**
Teach sqllogictest to respect tracing arguments and plumb them into the orchestrator. This allows clusterd to inherit user tracing env vars.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
Using this diff I was not able to prevent clusterd from logging INFO lines. Invocation was: `MZ_STARTUP_LOG_FILTER=warn MZ_LOG_FILTER=warn ./bin/sqllogictest`.

**Comment by benesch** [comment]
This looks about right to me. You'll want to delete `mz_ore::test::init_logging_default("warn")` too, I think?

File: src/sqllogictest/src/runner.rs [hunk 1]
```diff
@@ -959,9 +959,24 @@ impl<'a> RunnerInner<'a> {
             |_, _| PubSubClientConnection::noop(),
         );
         let persist_clients = Arc::new(persist_clients);
+        let (tracing_handle, _tracing_guard) = config
```
Comment (benesch) [inline src/sqllogictest/src/runner.rs:962]: You may need to stash this `_tracing_guard` somewhere to keep it alive as long as the server's alive? Maybe it's better to create the tracing guard in the main function, and pass the handle in here? Idk! But that might be what's causing this not to work.

Comment (mjibson) [reply src/sqllogictest/src/runner.rs]: Tried sticking it into a RefCell with no change.

**Comment by mjibson** [comment]
> You'll want to delete mz_ore::test::init_logging_default("warn") too, I think

That's already here. It panics without.

**Comment by benesch** [comment]
> Using this diff I was not able to prevent clusterd from logging INFO lines. Invocation was:
> 
> `MZ_STARTUP_LOG_FILTER=warn MZ_LOG_FILTER=warn ./bin/sqllogictest`

Oh, that's probably not the right invocation? I think you want `bin/sqllogictest --startup-log-filter=warn`. The `MZ_` prefix is specific to `environmentd`.

**Comment by mjibson** [comment]
> ./bin/sqllogictest -- --postgres-url $COCKROACH_URL test/sqllogictest/list.slt --startup-log-filter warn

Is the invocation I'm using and still no :(

**Review by benesch — APPROVED** [review]
Hammered on this. Should work now. Two tricks required;

  1. Set the `log_filter` system variable in the running environment to the value of the `--startup-log-filter` argument. Otherwise as soon as the environment is booted it stops respecting the startup log filter and uses the `log_filter` variable... which defaults to `info`.
  2. Plumb the `--startup-log-filter` through to `clusterd`, so that `clusterd` doesn't spam a few `info` logs on boot before it receives the current log filter from `environmentd`.

**Review by ParkMyCar — APPROVED** [review]

**Comment by mjibson** [comment]
This was failing with AlreadyReg errors from MetricsRegistry when running multiple files. I reverted the MetricsRegistry changes, so now the Tracing stuff gets its own registry. I'm not sure of the implications of this, but it does work now. Another option would probably be to move the tracing handle stuff back into runner.rs and stash the guard in the RunConfig again? This is so silly.

**Comment by benesch** [comment]
> This was failing with AlreadyReg errors from MetricsRegistry when running multiple files. I reverted the MetricsRegistry changes, so now the Tracing stuff gets its own registry. I'm not sure of the implications of this, but it does work now. Another option would probably be to move the tracing handle stuff back into runner.rs and stash the guard in the RunConfig again? This is so silly.

Ah, right, each run of the server needs a fresh metrics registry. Whatever. The metrics don't matter at all at the moment since this is a test. 🚢

**Comment by mjibson** [comment]
This is failing with

```
ERROR advance_ready{otel.name="execute"}: mz_adapter::coord::sequencer::inner: error while handling EXPLAIN statement: internal error: panic at the `local` optimization stage 
```

Also, run.log contains many

```
2023-10-07T06:15:13.616346Z  INFO mz_storage::storage_state: Applying configuration update...
```

which I thought should have gone away. Maybe slt is running in `info` (I believe the default) instead of `warn` (the previous hard-coded slt level which was removed in this PR). Probably we need to teach slt to again default to warn?

**Comment by benesch** [comment]
> Probably we need to teach slt to again default to warn?

Hmm, could be? I pushed up a commit that does that. I guess we'll see!

**Comment by benesch** [comment]
I have no idea. There are two different panics now. Sigh.

**Comment by def-** [comment]
As discussed on Slack (https://materializeinc.slack.com/archives/C01LKF361MZ/p1697838485777269) Those forced panics were there before, testing that optimizer is correctly converting them into errors. The way I was handling them in ci-logged-errors-detect (to find panic lines in our logs) was brittle and I just pushed a change inline to fix that.

**Comment by benesch** [comment]
Woohoo, glad to finally have this in. Thanks, @def-!

**Pull request merged**